### PR TITLE
Introduce xreceiveproxy for safer connextrouter

### PR DESCRIPTION
**Pull request by DaigaroCota**
This pull request addresses Item #-16 of the Three Sigma security audit.

### CODE REVIEW

**Comment by DaigaroCota** [comment]
Hey @brozorec , this pull request has a pattern refactor on how the xreceive function will work. 
I drafted the PR so that the changes are evident. 
I still have to update tests and such, however, I wanted to have your opinion on the effects of this changes.
I certainly now this will affect SDK, as the `to_` address in the xReceive call will change.
Please review also the attached issue and the discussion there.

File: packages/protocol/src/routers/XReceiveProxy.sol [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    bytes memory callData
+  )
+    external
+    returns (bytes memory)
+  {
+    IERC20 asset_ = IERC20(asset);
+
+    if (amount == 0) {
+      revert XReceiveProxy__xReceive_noValueTransfer();
+    }
```
Comment (brozorec) [inline packages/protocol/src/routers/XReceiveProxy.sol:49]: Why passing calldata only (without transfer of funds) wouldn't be a valid case?

File: packages/protocol/src/routers/ConnextRouter.sol [hunk 2]
```diff
@@ -137,31 +148,25 @@ contract ConnextRouter is BaseRouter, IXReceiver {
...
-    uint256 beforeSlipped;
-    if (amount > 0) {
-      // Ensure that at this entry point expected `asset` `amount` is received.
-      balance = IERC20(asset).balanceOf(address(this));
-      if (balance < amount) {
-        revert ConnextRouter__xReceive_notReceivedAssetBalance();
-      } else {
-        _tempTokenToCheck = Snapshot(asset, balance - amount);
-      }
+    IERC20 asset_ = IERC20(asset);
```
Comment (brozorec) [inline packages/protocol/src/routers/ConnextRouter.sol:156]: maybe no need to declare an additional variable here, use directly `IERC20(asset)` in the next 2 lines

File: packages/protocol/src/routers/ConnextRouter.sol [hunk 3]
```diff
@@ -82,6 +84,7 @@ contract ConnextRouter is BaseRouter, IXReceiver {
   IConnext public immutable connext;
 
   ConnextHandler public immutable handler;
+  address public immutable xreceiveProxy;
```
Comment (brozorec) [inline packages/protocol/src/routers/ConnextRouter.sol:87]: xReceiveProxy

**Review by brozorec — COMMENTED** [review]
That will require some changes in the SDK to replace the param `receiver` from ConnextRouter to the new `XReceiveProxy`. Est. time for that change: no more than 1d

File: packages/protocol/src/routers/XReceiveProxy.sol (hunk 1 above)
Comment (DaigaroCota) [reply packages/protocol/src/routers/XReceiveProxy.sol:49]: You are right, Updating it.

**Review by 0x73696d616f — APPROVED** [review]
Looks good

**Review by brozorec — APPROVED** [review]

**Pull request merged**